README.md
# Easy Http Errors

[![npm](https://img.shields.io/npm/v/easy-http-errors.svg?style=flat-square)](https://npmjs.org/package/easy-http-errors)
![node](https://img.shields.io/node/v/easy-http-errors.svg?style=flat-square)
[![Build Status](https://img.shields.io/travis/seegno/easy-http-errors/master.svg?style=flat-square)](https://travis-ci.org/seegno/easy-http-errors)
[![Coverage Status](https://img.shields.io/coveralls/seegno/easy-http-errors/master.svg?style=flat-square)](https://coveralls.io/github/seegno/easy-http-errors?branch=master)

A preset of HTTP errors that can be easily used to throw errors in your applications.

## Installation

### NPM

```sh
npm i easy-http-errors --save
```

### Yarn

```sh
yarn add easy-http-errors
```

## Usage

```js
// ES6 import.
import { BadRequestError } from 'easy-http-errors';

// Throw the default bad request.
throw new BadRequestError();

// Throw a bad request with a custom message and properties.
throw new BadRequestError('Ups, this is a bad request', { foo: 'bar' });
```

## List of errors

| Status Code | Name                               |
| :---------: | -----------------------------------|
| 400         | BadRequestError                    |
| 401         | UnauthorizedError                  |
| 402         | PaymentRequiredError               |
| 403         | ForbiddenError                     |
| 404         | NotFoundError                      |
| 405         | MethodNotAllowedError              |
| 406         | NotAcceptableError                 |
| 407         | ProxyAuthenticationRequiredError   |
| 408         | RequestTimeoutError                |
| 409         | ConflictError                      |
| 410         | GoneError                          |
| 411         | LengthRequiredError                |
| 412         | PreconditionFailedError            |
| 413         | PayloadTooLargeError               |
| 414         | URITooLongError                    |
| 415         | UnsupportedMediaTypeError          |
| 416         | RangeNotSatisfiableError           |
| 417         | ExpectationFailedError             |
| 418         | ImATeapotError                     |
| 421         | MisdirectedRequestError            |
| 422         | UnprocessableEntityError           |
| 423         | LockedError                        |
| 424         | FailedDependencyError              |
| 425         | UnorderedCollectionError           |
| 426         | UpgradeRequiredError               |
| 428         | PreconditionRequiredError          |
| 429         | TooManyRequestsError               |
| 431         | RequestHeaderFieldsTooLargeError   |
| 451         | UnavailableForLegalReasonsError    |
| 500         | InternalServerError                |
| 501         | NotImplementedError                |
| 502         | BadGatewayError                    |
| 503         | ServiceUnavailableError            |
| 504         | GatewayTimeoutError                |
| 505         | HTTPVersionNotSupportedError       |
| 506         | VariantAlsoNegotiatesError         |
| 507         | InsufficientStorageError           |
| 508         | LoopDetectedError                  |
| 509         | BandwidthLimitExceededError        |
| 510         | NotExtendedError                   |
| 511         | NetworkAuthenticationRequiredError |

## Tests

Run the tests from the root directory:

```sh
npm test
```

## Contributing & Development

### Contributing

Found a bug or want to suggest something? Take a look first on the current and closed [issues](https://github.com/seegno/easy-http-errors/issues). If it is something new, please [submit an issue](https://github.com/seegno/easy-http-errors/issues/new).

### Develop

It will be awesome if you can help us evolve `easy-http-errors`. Want to help?

1. [Fork it](https://github.com/seegno/easy-http-errors).
2. `npm install`.
3. Hack away.
4. Run the tests: `npm test`.
5. Create a [Pull Request](https://github.com/seegno/easy-http-errors/compare).


CHANGELOG.md
# Changelog

## [v2.0.0](https://github.com/seegno/easy-http-errors/tree/v2.0.0) (2019-03-19)
[Full Changelog](https://github.com/seegno/easy-http-errors/compare/v1.0.0...v2.0.0)

**Merged pull requests:**

- Fix NotAcceptable status code [\#5](https://github.com/seegno/easy-http-errors/pull/5) ([franciscomorais](https://github.com/franciscomorais))
- Drop node 4 support [\#4](https://github.com/seegno/easy-http-errors/pull/4) ([nunorafaelrocha](https://github.com/nunorafaelrocha))
- Update standard-http-error@2.0.1 [\#3](https://github.com/seegno/easy-http-errors/pull/3) ([nunorafaelrocha](https://github.com/nunorafaelrocha))
- Add coveralls [\#2](https://github.com/seegno/easy-http-errors/pull/2) ([nunorafaelrocha](https://github.com/nunorafaelrocha))

## [v1.0.0](https://github.com/seegno/easy-http-errors/tree/v1.0.0) (2017-07-28)
**Merged pull requests:**

- Add initial implementation [\#1](https://github.com/seegno/easy-http-errors/pull/1) ([nunorafaelrocha](https://github.com/nunorafaelrocha))


test/index.test.js

/**
 * Module dependencies.
 */

const StandardHttpError = require('standard-http-error');
const httpErrors = require('../');
const errors = require('../errors');

/**
 * Test `easy-http-errors`.
 */

describe('easy-http-errors', () => {
  const mappedErrors = errors.reduce((result, error) => {
    result[error.name] = error;

    return result;
  }, {});

  it('should create all http-errors', () => {
    for (const name in httpErrors) {
      try {
        throw new httpErrors[name]();
      } catch (e) {
        expect(e).toBeInstanceOf(StandardHttpError);
        expect(e.name).toEqual(mappedErrors[name].name);
        expect(e.code).toEqual(mappedErrors[name].status);
        expect(e.message).toEqual(mappedErrors[name].message);
      }
    }
  });

  it('should create errors with a custom message and properties', () => {
    for (const name in httpErrors) {
      try {
        throw new httpErrors[name]('foobar', { qux: 'biz' });
      } catch (e) {
        expect(e.code).toEqual(expect.any(Number));
        expect(e.message).toEqual('foobar');
        expect(e.qux).toEqual('biz');
      }
    }
  });
});


errors.js

/**
 * Export `errors`.
 */

module.exports = [
  { message: 'Bad request', name: 'BadRequestError', status: 400 },
  { message: 'Unauthorized', name: 'UnauthorizedError', status: 401 },
  { message: 'Payment required', name: 'PaymentRequiredError', status: 402 },
  { message: 'Forbidden', name: 'ForbiddenError', status: 403 },
  { message: 'Not Found', name: 'NotFoundError', status: 404 },
  { message: 'Method Not Allowed', name: 'MethodNotAllowedError', status: 405 },
  { message: 'Not Acceptable', name: 'NotAcceptableError', status: 406 },
  { message: 'Proxy Authentication Required', name: 'ProxyAuthenticationRequiredError', status: 407 },
  { message: 'Request Timeout', name: 'RequestTimeoutError', status: 408 },
  { message: 'Conflict', name: 'ConflictError', status: 409 },
  { message: 'Gone', name: 'GoneError', status: 410 },
  { message: 'Length Required', name: 'LengthRequiredError', status: 411 },
  { message: 'Precondition Failed', name: 'PreconditionFailedError', status: 412 },
  { message: 'Payload Too Large', name: 'PayloadTooLargeError', status: 413 },
  { message: 'URI Too Long', name: 'URITooLongError', status: 414 },
  { message: 'Unsupported Media Type', name: 'UnsupportedMediaTypeError', status: 415 },
  { message: 'Range Not Satisfiable', name: 'RangeNotSatisfiableError', status: 416 },
  { message: 'Expectation Failed', name: 'ExpectationFailedError', status: 417 },
  { message: 'I\'m A Tea pot', name: 'ImATeapotError', status: 418 },
  { message: 'Misdirected Request', name: 'MisdirectedRequestError', status: 421 },
  { message: 'Unprocessable Entity', name: 'UnprocessableEntityError', status: 422 },
  { message: 'Locked', name: 'LockedError', status: 423 },
  { message: 'Failed Dependency', name: 'FailedDependencyError', status: 424 },
  { message: 'Unordered Collection', name: 'UnorderedCollectionError', status: 425 },
  { message: 'Upgrade Required', name: 'UpgradeRequiredError', status: 426 },
  { message: 'Precondition Required', name: 'PreconditionRequiredError', status: 428 },
  { message: 'Too Many Requests', name: 'TooManyRequestsError', status: 429 },
  { message: 'Request Header Fields Too Large', name: 'RequestHeaderFieldsTooLargeError', status: 431 },
  { message: 'Unavailable For Legal Reasons', name: 'UnavailableForLegalReasonsError', status: 451 },
  { message: 'Internal Server Error', name: 'InternalServerError', status: 500 },
  { message: 'Not Implemented', name: 'NotImplementedError', status: 501 },
  { message: 'Bad Gateway', name: 'BadGatewayError', status: 502 },
  { message: 'Service Unavailable', name: 'ServiceUnavailableError', status: 503 },
  { message: 'Gateway Timeout', name: 'GatewayTimeoutError', status: 504 },
  { message: 'HTTP Version Not Supported', name: 'HTTPVersionNotSupportedError', status: 505 },
  { message: 'Variant Also Negotiates', name: 'VariantAlsoNegotiatesError', status: 506 },
  { message: 'Insufficient Storage', name: 'InsufficientStorageError', status: 507 },
  { message: 'Loop Detected', name: 'LoopDetectedError', status: 508 },
  { message: 'Bandwidth Limit Exceeded', name: 'BandwidthLimitExceededError', status: 509 },
  { message: 'Not Extended', name: 'NotExtendedError', status: 510 },
  { message: 'Network Authentication Required', name: 'NetworkAuthenticationRequiredError', status: 511 }
];
